File: modules/ls-index-model/src/ls/index/groups.scala
package ls.index

/** Reasons a rename group is rejected. Stored as a bitmask so the request
  * path is one integer test; expanded to messages only on rejection.
  */
object UnsafeReason:
  inline val External = 1L << 0
  inline val GeneratedOccurrence = 1L << 1
  inline val ReadonlyOccurrence = 1L << 2
  inline val OverrideFamily = 1L << 3
  inline val SyntheticOnly = 1L << 4
  inline val PcOnly = 1L << 5
  inline val SharedSourceDisagreement = 1L << 6
  inline val UnsupportedSymbolFamily = 1L << 7
  inline val DependencySource = 1L << 8

  def explain(mask: Long): List[String] =
    val msgs = List.newBuilder[String]
    if (mask & External) != 0 then
      msgs += "symbol is defined outside the workspace"
    if (mask & GeneratedOccurrence) != 0 then
      msgs += "symbol has occurrences in generated sources"
    if (mask & ReadonlyOccurrence) != 0 then
      msgs += "symbol has occurrences in readonly sources"
    if (mask & OverrideFamily) != 0 then
      msgs += "symbol participates in an override family that cannot be renamed safely"
    if (mask & SyntheticOnly) != 0 then
      msgs += "symbol only has synthetic occurrences"
    if (mask & PcOnly) != 0 then
      msgs += "symbol is provided by a PC-only plugin and is not present in fresh SemanticDB"
    if (mask & SharedSourceDisagreement) != 0 then
      msgs += "targets sharing this source disagree on the rename group"
    if (mask & UnsupportedSymbolFamily) != 0 then
      msgs += "symbol family (e.g. apply/unapply, exported symbol) is not safely renameable"
    if (mask & DependencySource) != 0 then
      msgs += "symbol has occurrences in dependency sources"
    msgs.result()

/** Precomputed at ingest; consulted at rename request time. */
final case class RenameProfile(
    isLocal: Boolean,
    isExternal: Boolean,
    hasGeneratedOccurrences: Boolean,
    hasReadonlyOccurrences: Boolean,
    hasOverrideFamily: Boolean,
    hasCompanion: Boolean,
    editableOccurrenceCount: Int,
    unsafeReasonMask: Long
):
  def isSafe: Boolean = unsafeReasonMask == 0L

object RenameProfile:
  val empty: RenameProfile =
    RenameProfile(false, false, false, false, false, false, 0, 0L)

File: crates/ls-index-model/src/groups.rs
//! The rename-safety profile precomputed per rename group at ingest.
//!
//! Mirrors the Scala `ls.index.RenameProfile`. The reason-bit constants and
//! their messages live in [`crate::unsafe_reason`]; this is the per-group record
//! the rename request path consults.

/// Precomputed at ingest; consulted at rename request time.
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct RenameProfile {
    pub is_local: bool,
    pub is_external: bool,
    pub has_generated_occurrences: bool,
    pub has_readonly_occurrences: bool,
    pub has_override_family: bool,
    pub has_companion: bool,
    pub editable_occurrence_count: u32,
    /// Reason mask (see [`crate::unsafe_reason`]); `0` means safe to rename.
    pub unsafe_reason_mask: u64,
}

impl RenameProfile {
    /// A rename group is safe exactly when no reason bit is set.
    #[inline]
    pub fn is_safe(&self) -> bool {
        self.unsafe_reason_mask == 0
    }

    /// The all-clear default (`RenameProfile.empty` in Scala).
    pub fn empty() -> Self {
        RenameProfile {
            is_local: false,
            is_external: false,
            has_generated_occurrences: false,
            has_readonly_occurrences: false,
            has_override_family: false,
            has_companion: false,
            editable_occurrence_count: 0,
            unsafe_reason_mask: 0,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_profile_is_safe() {
        let p = RenameProfile::empty();
        assert!(p.is_safe());
        assert_eq!(p.editable_occurrence_count, 0);
    }

    #[test]
    fn nonzero_mask_is_unsafe() {
        let mut p = RenameProfile::empty();
        p.unsafe_reason_mask = crate::unsafe_reason::EXTERNAL;
        assert!(!p.is_safe());
    }
}
